### PR TITLE
Fixed #33916 -- Added support for serialization of enum.Flag in migrations.

### PR DESCRIPTION
**Pull request by KommuSoft**

### CODE REVIEW

File: django/db/migrations/serializer.py [hunk 1]
```diff
@@ -154,7 +160,9 @@ def serialize(self):
                 "import %s" % module
             }
         # Further error checking
-        if self.value.__name__ == "<lambda>":
+        # use .co_name, since this can not be modified, whereas __name__ can be altered
+        # https://stackoverflow.com/a/56451124/67579
+        if self.value.__code__.co_name == "<lambda>":
```
Comment (felixxm) [inline django/db/migrations/serializer.py:165]: This is not related with `EnumSerializer` and cannot be proceeded in the same PR. Also, assigning lambdas to a variable and changing their `__name__` sounds really niche. I'm not sure if it's worth changing :thinking: and we use the same logic in `django.contrib.admin`.

Nevertheless, all bugfixes require [an accepted ticket](https://code.djangoproject.com/newticket), please create a new one.

File: django/db/migrations/serializer.py [hunk 2]
```diff
@@ -125,8 +125,14 @@ class EnumSerializer(BaseSerializer):
     def serialize(self):
         enum_class = self.value.__class__
         module = enum_class.__module__
+        if issubclass(enum_class, enum.Flag):
+            members, __ = enum._decompose(enum_class, self.value)
+        else:
+            members = (self.value,)
```
Comment (felixxm) [inline django/db/migrations/serializer.py:131]: In Python 3.11 it's possible to iterate over members instead of using a private API. Moreover `_decompose` is not available anymore (see https://github.com/python/cpython/commit/7aaeb2a3d682ecba125c33511e4b4796021d2f82):

```suggestion
        if issubclass(enum_class, enum.Flag):
            if PY311:
                members = list(self.value)
            else:
                members, _ = enum._decompose(enum_class, self.value)
                members = reversed(members)
        else:
            members = (self.value,)
```

File: django/db/migrations/serializer.py [hunk 3]
```diff
@@ -125,8 +125,14 @@ class EnumSerializer(BaseSerializer):
...
         return (
-            "%s.%s[%r]" % (module, enum_class.__qualname__, self.value.name),
+            " | ".join(
+                f"{module}.{enum_class.__qualname__}[{item.name!r}]" for item in members
+            ),
```
Comment (felixxm) [inline django/db/migrations/serializer.py:135]: A list comprehension is preferable here as `str.join()` converts to list internally anyway. It is better performance to provide a list up front.
 - https://stackoverflow.com/questions/9060653/list-comprehension-without-in-python/9061024#9061024
 - https://github.com/adamchainz/flake8-comprehensions/issues/156

```suggestion
            " | ".join([
                f"{module}.{enum_class.__qualname__}[{item.name!r}]" for item in members
            ]),
```

File: django/db/migrations/serializer.py (hunk 1 above)
Comment (felixxm) [reply django/db/migrations/serializer.py:165]:
```suggestion
        if self.value.__name__ == "<lambda>":
```

**Review by felixxm — COMMENTED** [review]
@KommuSoft Thanks :+1: I moved tests that work without any changes to a separate PR, see #15962. Please add also a release note and mention `enum.Flag` in ["Serializing values"](https://docs.djangoproject.com/en/stable/topics/migrations/#serializing-values), e.g.
```diff
diff --git a/docs/topics/migrations.txt b/docs/topics/migrations.txt
index cf3451a782..9c32fa9a53 100644
--- a/docs/topics/migrations.txt
+++ b/docs/topics/migrations.txt
@@ -739,7 +739,7 @@ Django can serialize the following:
 - ``datetime.date``, ``datetime.time``, and ``datetime.datetime`` instances
   (include those that are timezone-aware)
 - ``decimal.Decimal`` instances
-- ``enum.Enum`` instances
+- ``enum.Enum`` and ``enum.Flag`` instances
 - ``uuid.UUID`` instances
 - :func:`functools.partial` and :class:`functools.partialmethod` instances
   which have serializable ``func``, ``args``, and ``keywords`` values.
@@ -756,6 +756,10 @@ Django can serialize the following:
 - Any class reference (must be in module's top-level scope)
 - Anything with a custom ``deconstruct()`` method (:ref:`see below <custom-deconstruct-method>`)
 
+.. versionchanged:: 4.2
+
+    Serialization support for ``enum.Flag`` was added.
+
 Django cannot serialize:
 
 - Nested classes
```

File: django/db/migrations/serializer.py (hunk 2 above)
Comment (felixxm) [reply django/db/migrations/serializer.py:131]: `reversed` to get members in definition order, as in Python 3.11.

**Comment by KommuSoft** [comment]
Updated based on the PR comments.

**Comment by felixxm** [comment]
@KommuSoft Thanks :+1: Welcome aboard :boat: 

I added release notes and removed redundant tests, see #15996.
